templates/nginx/nginx-config.yaml

```
@@ -37,12 +37,48 @@ data:
       {{ tpl . $ | nindent 6 }}
       {{- end }}
 
+      {{- if .Values.alertmanager.enabled }}
+      upstream alertmanager {
+        server {{ template "cortex.fullname" . }}-alertmanager.{{ $rootDomain }};
+        keepalive {{ .Values.nginx.config.upstream_keepalive }};
+      }
+      {{- end }}
+
+      {{- if .Values.configs.enabled }}
+      upstream configs {
+        server {{ template "cortex.fullname" . }}-configs.{{ $rootDomain }};
+        keepalive {{ .Values.nginx.config.upstream_keepalive }};
+      }
+      {{- end }}
+
+      {{- if .Values.distributor.enabled }}
+      upstream distributor {
+        server {{ template "cortex.fullname" . }}-distributor.{{ $rootDomain }};
```

Based on the definition here: http://nginx.org/en/docs/http/ngx_http_upstream_module.html#server
> A domain name that resolves to several IP addresses defines multiple servers at once.

I think it would be better to use the `distributor-headless` service. We should also use `resolve` to monitor for changes in the set of pods.
> `resolve` monitors changes of the IP addresses that correspond to a domain name of the server, and automatically modifies the upstream configuration without the need of restarting nginx
```suggestion
        server {{ template "cortex.fullname" . }}-distributor-headless.{{ $rootDomain }} resolve;
```

---

I guess we should do that for every service then that has a headless service.

---

The `resolve` option is only supported for nginx commercial, so not sure we can do this.

---

> > The `resolve` option is only supported for nginx commercial, so not sure we can do this.
> 
> Oh, maybe that's why ingress-nginx uses lua to resolve pod addresses. This is convincing me that the nginx gateway in this chart isn't very useful and most use cases might be better served by Ingress.

Agree. However we would have to abandon some features

---

Okay, so without `resolve`, how can we distribute requests... Is the design assumption that the `keepalive 16` will result in open connections roughly balanced to each pod? And nginx will round-robin requests to those connections?

BTW, here I use linkerd to balance load to the distributors. It works quite well for that, but I wouldn't expect everyone to deploy linkerd.

---

> Okay, so without `resolve`, how can we distribute requests... Is the design assumption that the `keepalive 16` will result in open connections roughly balanced to each pod? And nginx will round-robin requests to those connections?

I would assume that we actually have to do the opposite and ensure that we do not use any kind of keepalive connection to achieve fair distribution.

https://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout setting this to `0` should disable that completely

> The `resolve` option is only supported for nginx commercial, so not sure we can do this.

Where does it say that?